### ISSUE TITLE
Print Current Proof State when a Command or Tactic Fails

### ISSUE THREAD

**Issue by gmalecha** [coq/coq#19601]
### Is your feature request related to a problem?

When running large CI jobs, it can be very useful to get more information about a failure that simply "tactic failed". For example,

```coq
    Goal True.
    Proof.
      rewrite Nat.add_0_l.
```

produces the error

```coq
Error: Ltac call to "rewrite (ssrrwargs) (ssrclauses)" failed.
       The LHS of Nat.add_0_l
    (0 + _)
does not match any subterm of the goal
```

But it would be useful to know what the actual goal is without needing to start evaluating the file.

This might be especially useful for larger tactics and when automation because in cases like that, you might have a lot of proof breakage that is very similar and looking at a CI log can give you a quick idea of what the problem might be.

### Proposed solution

When running a command the causes `coqc` to fail, print the current proof state if there is a corresponding proof state.
Additionally, it might be useful to print the actual tactic or command that fails.

**Comment by ejgallego** [comment]
Seems indeed pretty useful to me. I guess implementing this via a `coqc` flag would work for you right folks?

Something to consider is how this flag would relate to open issues about verbosity of printing and silent builds.

**Note:** in `fcc`, this kind of use cases are done with a plugin, which could allow users to do way more custom stuff (like try some proof repair, etc...)

**Comment by gmalecha** [comment]
> Seems indeed pretty useful to me. I guess implementing this via a coqc flag would work for you right folks?
A flag to `coqc` is indeed what I was thinking about. I don't know much about this sort of thing, but could do some hacking on it.

> Something to consider is how this flag would relate to open issues about verbosity of printing and silent builds.
Which issues are you thinking about? If it is an opt-in feature, then it seems like it would probably be compatible with other things, but this feature should not change anything about successful builds since it only prints when there is an error.

> Note: in `fcc`, this kind of use cases are done with a plugin, which could allow users to do way more custom stuff (like try some proof repair, etc...)
That could be a nice feature. For example, if you could instrument every interaction with its output (regardless of whether it succeeds or not), then you could generate proof movies without the need for a separate build. Perhaps the flag could be `--record-output=never|always|on-failure` or something like that?

A full plugin architecture for intercepting commands like this is probably a larger task than just this sort of flag. I might have cycles to contribute to a flag, but I don't think I have the cycles to do a plugin architecture.

**Comment by ejgallego** [comment]
> Which issues are you thinking about? If it is an opt-in feature, then it seems like it would probably be compatible with other things, but this feature should not change anything about successful builds since it only prints when there is an error.

Yes, I meant actual github issues that are already open, not actual "issues" sorry. My comment was mostly meant for other Coq devs as a reorg of Coq printing could happen.

> A flag to coqc is indeed what I was thinking about. I don't know much about this sort of thing, but could do some hacking on it.

That would be great! The place to tweak is https://github.com/coq/coq/blob/fe26a718a5f175e9720632b9fec8b797f476d139/toplevel/coqc.ml#L58 , however, at that point, you don't know what is the state to print the goals, as the exception with the error won't carry it.

I guess we could wrap it in `Ccompile`, or just make `Ccompile` return a `Result.t`.

> That could be a nice feature. For example, if you could instrument every interaction with its output (regardless of whether it succeeds or not), then you could generate proof movies without the need for a separate build. Perhaps the flag could be --record-output=never|always|on-failure or something like that?
>
> A full plugin architecture for intercepting commands like this is probably a larger task than just this sort of flag. I might have cycles to contribute to a flag, but I don't think I have the cycles to do a plugin architecture.

Indeed! Note that I was talking about `fcc`, the "Flèche Coq Compiler", which is a separate (for now) project.

`fcc` already has a plugin API, and a plugin example that records all goals: https://github.com/ejgallego/coq-lsp/tree/main/plugins/goaldump 

Your use case suggested a new plugin, or mode for the `goaldump` plugin.

**Comment by SkySkimmer** [comment]
>That would be great! The place to tweak is 

I don't know about that, I was imagining that coqc sets an option and the option changes some error handling at a lower level (probably in vernac). For instance it could make errors gets wrapped with `exception AlsoPrintGoal of goal * exn`.

**Comment by SkySkimmer** [comment]
It may even be needed to handle it at the ltac level to work with ltac `;`. 
I mean if you have `intros ...; destruct ...; failing_tactic` do you want to print the goal at the beginning of the command or the goal in the branch of the destruct which failed?
